### PR TITLE
Build Spark from source

### PR DESCRIPTION
**Pull request by razvan**
# Description

:warning:  CI: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/spark-k8s-operator-it-custom/132/

```
--- FAIL: kuttl (1905.79s)
    --- FAIL: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/delta-lake_openshift-false_spark-delta-lake-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source_delta-3.1.0 (209.63s)
        --- PASS: kuttl/harness/smoke_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source_s3-use-tls-true (156.61s)
        --- PASS: kuttl/harness/pyspark-ny-public-s3-image_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source_ny-tlc-report-0.1.0 (128.42s)
        --- PASS: kuttl/harness/spark-pi-private-s3_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source (87.89s)
        --- PASS: kuttl/harness/logging_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source_ny-tlc-report-0.1.0 (284.48s)
        --- PASS: kuttl/harness/pod_overrides_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source (114.80s)
        --- FAIL: kuttl/harness/iceberg_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source (300.95s)
        --- PASS: kuttl/harness/pyspark-ny-public-s3_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source (102.34s)
        --- PASS: kuttl/harness/spark-pi-public-s3_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source (76.81s)
        --- PASS: kuttl/harness/spark-examples_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source (53.22s)
        --- PASS: kuttl/harness/spark-history-server_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source_s3-use-tls-true (205.42s)
        --- PASS: kuttl/harness/spark-ny-public-s3_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source_s3-use-tls-true (137.36s)
        --- PASS: kuttl/harness/resources_openshift-false_spark-3.5.1,docker.stackable.tech_sandbox_spark-k8s_3.5.1-from-source (47.33s)
FAIL
ERROR:root:kuttl failed
```
## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
- [ ] Changes are OpenShift compatible
- [ ] All added packages (via microdnf or otherwise) have a comment on why they are added
- [ ] Things not downloaded from Red Hat repositories should be mirrored in the Stackable repository and downloaded from there
- [ ] All packages should have (if available) signatures/hashes verified
- [ ] Does your change affect an SBOM? Make sure to update all SBOMs
- [ ] Add an entry to the CHANGELOG.md file
- [ ] Integration tests ran successfully
```

<details>
<summary>TIP: Running integration tests with a new product image</summary>

The image can be built and uploaded to the kind cluster with the following commands:

```shell
bake --product <product> --image-version <stackable-image-version>
kind load docker-image <image-tagged-with-the-major-version> --name=<name-of-your-test-cluster>
```

See the output of `bake` to retrieve the image tag for `<image-tagged-with-the-major-version>`.
</details>

### CODE REVIEW

File: spark-k8s/Dockerfile [hunk 1]
```diff
@@ -18,9 +17,80 @@ ARG JACKSON_DATAFORMAT_XML
 ARG STAX2_API
 ARG WOODSTOX_CORE
 ARG JMX_EXPORTER
+
+WORKDIR /stackable
+
+# >>> Build spark
+# Compiling the tests takes a lot of time, so we skip them
+# -Dmaven.test.skip=true skips both the compilation and execution of tests
```
Comment (sbernauer) [inline spark-k8s/Dockerfile:25]: TIL :)

**Review by sbernauer — CHANGES_REQUESTED** [review]
Mostly Nits

File: spark-k8s/Dockerfile [hunk 2]
```diff
@@ -34,104 +104,47 @@ RUN microdnf update && \
...
+COPY --chown=stackable:stackable --from=builder /stackable/jmx /stackable/jmx
 
-RUN chmod +x /usr/bin/tini
+RUN ln -s "/stackable/jmx/jmx_prometheus_javaagent-${JMX_EXPORTER}.jar" /stackable/jmx/jmx_prometheus_javaagent.jar \
+    # Symlink example jar, so that we can easily use it in tests
+    && ln -s /stackable/spark/examples/jars/spark-examples_*.jar /stackable/spark/examples/jars/spark-examples.jar
 
-RUN chown -R stackable:stackable /stackable
+# Get the correct `tini` binary for our architecture.
+RUN curl --fail -L -o /usr/bin/tini "https://repo.stackable.tech/repository/packages/tini/tini-${TINI}-${TARGETARCH}" \
```
[the remark this answers is not in the document]
Comment (lfrancke) [reply spark-k8s/Dockerfile:139]: Personally I also find it nicer to have everything prepared in the builder and just copied over in the final one.
But this is something that I don't care too much about.

**Review by lfrancke — CHANGES_REQUESTED** [review]

File: spark-k8s/Dockerfile [hunk 3]
```diff
@@ -18,9 +17,80 @@ ARG JACKSON_DATAFORMAT_XML
...
+      -Dhadoop.version="$HADOOP_LONG_VERSION" \
+      -Dmaven.test.skip=true \
+      -DskipTests \
+      -P'hadoop-3' -Pkubernetes -Phive -Phive-thriftserver
+# <<< Build spark
+
+# We download these under dist so that log4shell checks them
+WORKDIR /stackable/spark-${PRODUCT}/dist/jars
+
+# Download aws module for Hadoop (support for s3a://)
```
[the remark this answers is not in the document]
Comment (razvan) [reply spark-k8s/Dockerfile]: done https://github.com/stackabletech/docker-images/pull/679/commits/2197bb387d1ac14337e71f2bc532281085f05637

File: spark-k8s/Dockerfile [hunk 4]
```diff
@@ -18,9 +17,80 @@ ARG JACKSON_DATAFORMAT_XML
...
+
+WORKDIR /stackable/jmx
+
+RUN curl --fail -L -O "https://repo.stackable.tech/repository/packages/jmx-exporter/jmx_prometheus_javaagent-${JMX_EXPORTER}.jar" 
+
+# ===
+# Mitigation for CVE-2021-44228 (Log4Shell)
+# This variable is supported as of Log4j version 2.10 and
+# disables the vulnerable feature
+ENV LOG4J_FORMAT_MSG_NO_LOOKUPS=true
```
[the remark this answers is not in the document]
Comment (razvan) [reply spark-k8s/Dockerfile]: https://github.com/stackabletech/docker-images/pull/679/commits/2197bb387d1ac14337e71f2bc532281085f05637

File: spark-k8s/Dockerfile [hunk 5]
```diff
@@ -18,9 +17,80 @@ ARG JACKSON_DATAFORMAT_XML
...
+# This will currently report vulnerable files being present, as it also alerts on
+# SocketNode.class, which we do not remove with our scripts.
+# Further investigation will be needed whether this should also be removed.
+COPY shared/log4shell_1.6.1-log4shell_Linux_x86_64 /bin/log4shell_scanner_x86_64
+COPY shared/log4shell_1.6.1-log4shell_Linux_aarch64 /bin/log4shell_scanner_aarch64
+COPY shared/log4shell_scanner /bin/log4shell_scanner
+RUN /bin/log4shell_scanner s /stackable/spark-${PRODUCT}/dist
+# ===
+
+FROM stackable/image/java-base
```
[the remark this answers is not in the document]
Comment (razvan) [reply spark-k8s/Dockerfile]: https://github.com/stackabletech/docker-images/pull/679/commits/2197bb387d1ac14337e71f2bc532281085f05637

File: spark-k8s/Dockerfile [hunk 6]
```diff
@@ -34,104 +104,47 @@ RUN microdnf update && \
...
     gzip \
     hostname \
-    # required for jmap and jps commands which are used by spark UI to show java processes
-    java-"${JAVA_VERSION}"-openjdk-devel \
     # required for spark startup scripts
     procps \
     python${PYTHON} \
     python${PYTHON}-pip \
+    # tar is required for `kubectl cp` which can be used to copy the log files or profiler flamegraph from the Pod
     tar \
```
[the remark this answers is not in the document]
Comment (razvan) [reply spark-k8s/Dockerfile]: https://github.com/stackabletech/docker-images/pull/679/commits/2197bb387d1ac14337e71f2bc532281085f05637

File: spark-k8s/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (razvan) [reply spark-k8s/Dockerfile]: https://github.com/stackabletech/docker-images/pull/679/commits/2197bb387d1ac14337e71f2bc532281085f05637

File: spark-k8s/Dockerfile [hunk 7]
```diff
@@ -2,14 +2,13 @@
 
 # Ignoring DL3038 globally because set `assumeyes=True` in dnf.conf in our base image
 # Ignoring DL4006 globally because we inherit the SHELL from our base image
-# hadolint global ignore=DL3038,DL4006
+# Ignoring DL3003 globally because we avoid unnecessary WORKDIR usages
+# Ignoring DL3006 globally because we don't want to pin java-* images
```
[the remark this answers is not in the document]
Comment (razvan) [reply spark-k8s/Dockerfile]: https://github.com/stackabletech/docker-images/pull/679/commits/2197bb387d1ac14337e71f2bc532281085f05637

**Review by lfrancke — APPROVED** [review]

File: spark-k8s/upload_new_spark_version.sh [hunk 8]
```diff
@@ -8,58 +8,58 @@ echo ""
...
+	rm -rf "$WORK_DIR"
 }
 
 # register the cleanup function to be called on the EXIT signal
 trap cleanup EXIT
 
 cd "$WORK_DIR" || exit
 
-bin_file="spark-${VERSION}-bin-hadoop3.tgz"
+src_file="spark-${VERSION}.tgz"
```
[the remark this answers is not in the document]
Comment (sbernauer) [reply spark-k8s/upload_new_spark_version.sh:32]: I have commented it on other PRs as well :) An other PRs kept the binaries. Chaos! :)

**Review by sbernauer — APPROVED** [review]
Thanks

**Pull request merged**